File: rd-net/RdFramework/Impl/AsyncRdMap.cs
using System;
using System.Collections;
using System.Collections.Generic;
using JetBrains.Annotations;
using JetBrains.Collections.Viewable;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.Rd.Base;
using JetBrains.Rd.Util;
using JetBrains.Serialization;

namespace JetBrains.Rd.Impl;

public class AsyncRdMap<K, V> : IRdBindable, IAsyncSource<MapEvent<K, V>>, IDictionary<K, V> where K : notnull
{
  private readonly RdMapBackend myMap;
  private readonly AsyncSignal<MapEvent<K, V>> mySignal = new();

  public AsyncRdMap(CtxReadDelegate<K> readKey, CtxWriteDelegate<K> writeKey, CtxReadDelegate<V> readValue, CtxWriteDelegate<V> writeValue)
  {
    myMap = new RdMapBackend(readKey, writeKey, readValue, writeValue);
    myMap.Change.Advise(Lifetime.Eternal, x => mySignal.Fire(x));
  }

  public IAsyncSource<MapEvent<K, V>> Change => mySignal;
  
  public bool IsMaster = false;
  
  public RdId RdId
  {
    get
    {
      lock (myMap)
        return myMap.RdId;
    }
    set
    {
      lock (myMap)
        myMap.RdId = value;
    }
  }

  public void PreBind(Lifetime lf, IRdDynamic parent, string name)
  {
    lock (myMap) 
      myMap.PreBind(lf, parent, name);
  }

  public void Bind()
  {
    lock (myMap) 
      myMap.Bind();
  }

  public void Identify(IIdentities identities, RdId id)
  {
    lock (myMap) 
      myMap.Identify(identities, id);
  }

  public bool OptimizeNested
  {
    get => true;
    set { }
  }
  
  public bool ValueCanBeNull
  {
    get => myMap.ValueCanBeNull;
    set => myMap.ValueCanBeNull = value;
  }

  public bool Async
  {
    get => true;
    set { }
  }

  public RName Location
  {
    get
    {
      lock (myMap)
        return myMap.Location;
    }
  }

  public IProtocol? TryGetProto()
  {
    return myMap.TryGetProto();
  }

  public bool TryGetSerializationContext(out SerializationCtx ctx)
  {
    return myMap.TryGetSerializationContext(out ctx);
  }

  public bool ContainsKey(K key)
  {
    return myMap.ContainsKey(key);
  }

  public void Add(K key, V value)
  {
    lock (myMap)
      myMap.Add(key, value);
  }

  public bool Remove(K key)
  {
    lock (myMap)
      return myMap.Remove(key);
  }

  public bool TryGetValue(K key, out V value)
  {
    return myMap.TryGetValue(key, out value);
  }

  public V this[K key]
  {
    get
    {
      lock (myMap)
        return myMap[key];
    }
    set
    {
      lock (myMap)
        myMap[key] = value; 
    }
  }

  public ICollection<K> Keys => myMap.Keys;
  public ICollection<V> Values => myMap.Values;

  void ICollection<KeyValuePair<K, V>>.Add(KeyValuePair<K, V> item) => Add(item.Key, item.Value);

  public void Clear()
  {
    lock (myMap) 
      myMap.Clear();
  }

  public bool Contains(KeyValuePair<K, V> item) => myMap.Contains(item);

  public void CopyTo(KeyValuePair<K, V>[] array, int arrayIndex)
  {
    myMap.CopyTo(array, arrayIndex);
  }

  public bool Remove(KeyValuePair<K, V> item)
  {
    lock (myMap)
      return myMap.Remove(item);
  }

  public int Count => myMap.Count;
  public bool IsReadOnly => myMap.IsReadOnly;
  
  public IEnumerator<KeyValuePair<K, V>> GetEnumerator() => myMap.GetEnumerator();
  IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

  [PublicAPI]
  public static AsyncRdMap<K, V> Read(SerializationCtx ctx, UnsafeReader reader)
  {
    return Read(ctx, reader, Polymorphic<K>.Read, Polymorphic<K>.Write, Polymorphic<V>.Read, Polymorphic<V>.Write);
  }
  
  [PublicAPI]
  public static AsyncRdMap<K,V> Read(SerializationCtx ctx, UnsafeReader reader, CtxReadDelegate<K> readKey, CtxWriteDelegate<K> writeKey, CtxReadDelegate<V> readValue, CtxWriteDelegate<V> writeValue)
  {
    var id = reader.ReadRdId();
    return new AsyncRdMap<K, V>(readKey, writeKey, readValue, writeValue).WithId(id);
  }

  [PublicAPI]
  public static void Write(SerializationCtx ctx, UnsafeWriter writer, AsyncRdMap<K, V> value)
  {
    Assertion.Assert(!value.RdId.IsNil);
    writer.Write(value.RdId);
  }


  [PublicAPI] public CtxReadDelegate<K> ReadKeyDelegate => myMap.ReadKeyDelegate;
  [PublicAPI] public CtxWriteDelegate<K> WriteKeyDelegate => myMap.WriteKeyDelegate;

  [PublicAPI] public CtxReadDelegate<V> ReadValueDelegate => myMap.ReadValueDelegate;
  [PublicAPI] public CtxWriteDelegate<V> WriteValueDelegate => myMap.WriteValueDelegate;

  private class RdMapBackend : RdMap<K, V>
  {
    public RdMapBackend(CtxReadDelegate<K> readKey, CtxWriteDelegate<K> writeKey, CtxReadDelegate<V> readValue, CtxWriteDelegate<V> writeValue) : base(readKey, writeKey, readValue, writeValue)
    {
      OptimizeNested = true;
      Async = true;
    }
    
    protected override void AssertBindingThread()
    {
    }

    protected override void AssertThreading()
    {
      
    }

    public override void OnWireReceived(IProtocol proto, SerializationCtx ctx, UnsafeReader stream, IRdWireableDispatchHelper dispatchHelper)
    {
      lock (this)
        base.OnWireReceived(proto, ctx, stream, new DelegatingDispatchHelper(dispatchHelper));
    }
  }
  
  private class DelegatingDispatchHelper : IRdWireableDispatchHelper
  {
    private readonly IRdWireableDispatchHelper myDispatchHelper;
    
    public RdId RdId => myDispatchHelper.RdId;
    public Lifetime Lifetime => myDispatchHelper.Lifetime;

    public DelegatingDispatchHelper(IRdWireableDispatchHelper dispatchHelper)
    {
      myDispatchHelper = dispatchHelper;
    }
    
    public void Dispatch(Lifetime lifetime, IScheduler? scheduler, Action action)
    {
      myDispatchHelper.Dispatch(SynchronousScheduler.Instance, action);
    }
  }

  public void Print(PrettyPrinter printer)
  {
    lock (myMap) 
      myMap.Print(printer);
  }

  public void AdviseOn(Lifetime lifetime, IScheduler scheduler, Action<MapEvent<K, V>> action)
  {
    lock (myMap)
    {
      myMap.Advise(lifetime, e =>
      {
        scheduler.Queue(() => action(e));
      });
    }
  }
}

File: rd-kt/rd-framework/src/main/kotlin/com/jetbrains/rd/framework/impl/AsyncRdMap.kt
package com.jetbrains.rd.framework.impl

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.framework.base.*
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.reactive.IScheduler
import com.jetbrains.rd.util.reactive.IViewableMap
import com.jetbrains.rd.util.string.RName
import com.jetbrains.rd.util.threading.SynchronousScheduler
import java.util.function.BiFunction
import java.util.function.Function

class AsyncRdMap<K : Any, V : Any> private constructor(
    private val map: BackendRdMap<K, V>
) : IRdBindable,  MutableMap<K, V>, IAsyncSource2<IViewableMap.Event<K, V>> {

    constructor(keySzr: ISerializer<K> = Polymorphic(), valSzr: ISerializer<V> = Polymorphic(),) : this(BackendRdMap(keySzr, valSzr))

    companion object : ISerializer<AsyncRdMap<*, *>> {
        fun<K:Any, V:Any> read(ctx: SerializationCtx, buffer: AbstractBuffer, keySzr: ISerializer<K>, valSzr: ISerializer<V>): AsyncRdMap<K, V> = AsyncRdMap(keySzr, valSzr).withId(RdId.read(buffer))
        override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: AsyncRdMap<*, *>) = value.rdid.write(buffer)

        override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): AsyncRdMap<*, *> {
            return read(ctx, buffer, Polymorphic(), Polymorphic())
        }
    }

    val change: IAsyncSource2<IViewableMap.Event<K, V>> = object : IAsyncSource2<IViewableMap.Event<K, V>> {
        override fun adviseOn(lifetime: Lifetime, scheduler: IScheduler, handler: (IViewableMap.Event<K, V>) -> Unit) {
            map.change.advise(lifetime) {
                scheduler.queue { handler(it) }
            }
        }
    }

    override fun adviseOn(lifetime: Lifetime, scheduler: IScheduler, handler: (IViewableMap.Event<K, V>) -> Unit) {
        synchronized(map) {
            map.advise(lifetime) { event ->
                scheduler.queue { handler(event) }
            }
        }
    }

    fun withId(id: RdId) : AsyncRdMap<K, V> {
        synchronized(map ) {
            map.withId(id)
        }
        return this
    }

    var optimizeNested: Boolean
        get() = true
        set(value) {}

    var async: Boolean
        get() = true
        set(value) {}

    override val rdid: RdId
        get() = synchronized(map) {
            map.rdid
        }

    override fun preBind(lf: Lifetime, parent: IRdDynamic, name: String) {
        synchronized(map) {
            map.preBind(lf, parent, name)
        }
    }

    override fun bind() {
        synchronized(map) {
            map.bind()
        }
    }

    override fun identify(identities: IIdentities, id: RdId) {
        synchronized(map) {
            map.identify(identities, id)
        }
    }

    override fun deepClone(): IRdBindable {
        synchronized(map) {
            val clone = map.deepClone()
            return AsyncRdMap(clone)
        }
    }

    override val protocol: IProtocol?
        get() = map.protocol

    override val serializationContext: SerializationCtx?
        get() = map.serializationContext

    override val location: RName
        get() = map.location

    override val size: Int
        get() = map.size

    override fun containsKey(key: K): Boolean {
        return map.containsKey(key)
    }

    override fun containsValue(value: V): Boolean {
        return map.containsValue(value)
    }

    override val entries: MutableSet<MutableMap.MutableEntry<K, V>>
        get() = map.entries // todo return wrapper to synchronize modification operations

    override val keys: MutableSet<K>
        get() = map.keys  // todo return wrapper to synchronize modification operations

    override val values: MutableCollection<V>
        get() = map.values  // todo return wrapper to synchronize modification operations

    override fun clear() {
        synchronized(map) {
            map.clear()
        }
    }

    override fun compute(key: K, remappingFunction: BiFunction<in K, in V?, out V?>): V? {
        TODO()
    }

    override fun computeIfAbsent(key: K, mappingFunction: Function<in K, out V>): V {
        TODO()
    }

    override fun computeIfPresent(key: K, remappingFunction: BiFunction<in K, in V, out V?>): V? {
        TODO()
    }

    override fun get(key: K): V? = map[key]

    override fun getOrDefault(key: K, defaultValue: V): V {
        return map.getOrDefault(key,defaultValue)
    }

    override fun isEmpty(): Boolean {
        return map.isEmpty()
    }

    override fun merge(key: K, value: V, remappingFunction: BiFunction<in V, in V, out V?>): V? {
        TODO()
    }

    override fun put(key: K, value: V): V? {
        return synchronized(map) {
            map.put(key, value)
        }
    }

    override fun putAll(from: Map<out K, V>) {
        synchronized(map) {
            map.putAll(from)
        }
    }

    override fun putIfAbsent(key: K, value: V): V? {
        return synchronized(map) {
            map.putIfAbsent(key, value)
        }
    }

    override fun remove(key: K): V? {
        return synchronized(map) {
            map.remove(key)
        }
    }

    override fun remove(key: K, value: V): Boolean {
        return synchronized(map) {
            map.remove(key, value)
        }
    }

    override fun replace(key: K, oldValue: V, newValue: V): Boolean {
        return synchronized(map) {
            map.replace(key, oldValue, newValue)
        }
    }

    override fun replace(key: K, value: V): V? {
        return synchronized(map) {
            map.replace(key, value)
        }
    }

    override fun replaceAll(function: BiFunction<in K, in V, out V>) {
        TODO()
    }

    private class BackendRdMap<K : Any, V : Any>(
        keySzr: ISerializer<K> = Polymorphic(),
        valSzr: ISerializer<V> = Polymorphic(),
    ) : RdMap<K, V>(keySzr, valSzr) {

        init {
            async = true
            optimizeNested = true
        }

        override fun assertBindingThread() {
        }

        override fun assertThreading() {
        }

        override fun onWireReceived(proto: IProtocol, buffer: AbstractBuffer, ctx: SerializationCtx, dispatchHelper: IRdWireableDispatchHelper) {
            synchronized(this) {
                super.onWireReceived(proto, buffer, ctx, object : IRdWireableDispatchHelper {
                    override val rdId: RdId
                        get() = dispatchHelper.rdId
                    override val lifetime: Lifetime
                        get() = dispatchHelper.lifetime

                    override fun dispatch(lifetime: Lifetime, scheduler: IScheduler?, action: () -> Unit) {
                        dispatchHelper.dispatch(lifetime, SynchronousScheduler, action)
                    }
                })
            }
        }

        override fun deepClone(): BackendRdMap<K, V> = BackendRdMap(keySzr, valSzr).also { for ((k,v) in it) { it[k] = v.deepClonePolymorphic() } }
    }
}